### PR TITLE
Expands x86 decoder to recognise 80386 opcodes.

### PR DESCRIPTION
**Pull request by TomHarte**

### CODE REVIEW

**Comment by TomHarte** [comment]
If it helps for context, I'm working towards full 80386 decoding, with my primary source being Programming the 80386 by Crawford and Gelsinger. I had previously implemented 8086 decoding using The 8086 Book by Rector and Alexy.

So: what I think is present on an 80286 is deduced from much worse sources than the two main processors on either side of it, and I don't intend to land this until the entire 80386 is in there at which point hopefully whatever is or isn't supposed to be available on an 80286 will be easy to include or exclude from that model.

That all being said, per the diagram I'm using (from the 80386 book):
* 0xf1 and 0xd6 aren't mapped to anything;
* 0x6x look 386 specific; and
* 0xd1 and 0xd3 look like definite omissions.

The immediate plan is to extend the 8086 decoder test cases both with targetted tests for what's added at each generation of processor and with a bunch more of the randomly-generated disassemblies but against the wider instruction set.

There are also some changes to `Instruction` coming, since it need to be able to specify more, but I think I'm going to give it a size field and pack things so that you can pack them into memory more tightly, e.g. omitting the field for an immediate value if there isn't one.

I'll worry about execution much later. The 80386 book contains C pseudocode for every instruction, and similar explanations are available online, so I expect that to be a slog but not problematic.

**Comment by TomHarte** [comment]
> I suspect the case in this define should end with a break or return statement

Yeah, I think that's correct. There are also a few other issues lurking, I think. I'm working on the unit tests now, so hopefully the next push will clean things up a lot.

File: InstructionSets/x86/DataPointerResolver.hpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+		template <bool is_write, typename DataT> DataT access(
+			RegistersT &registers,
+			MemoryT &memory,
+			const Instruction<is_32bit(model)> &instruction,
+			DataPointer pointer,
+			decltype(RegistersT::eSP) memory_mask,
+			DataT &value) {
+				const Source source = pointer.source();
+
+#define read_or_write(v, x, allow_write)	\
```
Comment (PatrickvL) [inline InstructionSets/x86/DataPointerResolver.hpp]: Perhaps replace `allow_write` by `is_not_indirect` (or invert that name, the use and the given argument values) to make it a bit more clear why this argument is switching from a write to a read?

**Comment by TomHarte** [comment]
> How do these arguments compile?

I'm not sure I fully understand the question; if it helps at all then currently the only consumer of the DataPointResolver is x86DataPointResolverTests.mm, part of the Xcode-hosted test suite. It currently has exactly one test, that being sufficient to cross the compilation threshold.

It has struck me since this morning that I actually need to factor out the address calculation for the benefit of `LEA`, but that should be no big deal. It also doesn't technically need to be correct on the details for the time being, as long as it proves that the necessary inputs are retained and establishes who has responsibility for which part of interpreting them.

At some point when I'm further along with the decoder I guess I can find an x86 machine and put together some LEA-specific unit tests*. It'd probably be smart to have at least one instruction implemented before I call this seemingly done, again just to verify wider implementation decisions.

\* incredibly, the DJGPP port of GCC to DOS still seems to be active. So you can compile C++17 with inline assembly. That probably makes sense for collecting 32-bit test data. I'm sure I can cope with Turbo C or something like that long enough to throw together some 16-bit tests. No need to get myself heavily invested in jumping in and out of protected and real mode right now — I'm sure that'd be 90% of the task if I let it.

**Comment by TomHarte** [comment]
Cool, thanks for all the review! All fixed, at least for now.

**Comment by qeeg** [comment]
8ch is a valid opcode on 8086/8088 as well, and does exactly what it does on later processors.

**Comment by TomHarte** [comment]
> 8ch is a valid opcode on 8086/8088 as well, and does exactly what it does on later processors.

Indeed, I had completely misread my source. At least twice.

**Pull request merged**